### PR TITLE
Add classpath + aggregate dependency information to minimal project buil...

### PR DESCRIPTION
**Pull request by jsuereth**
...d.

Replaces #286

A few caveats

* the 'configuration' exposed is Ivy's string-y format.  Ideally we should hide this behind somethig more typesafe.
  To understand what it means, you have to fully understand configuration mappings in Ivy.
* We've put the dependency listing as an *optional* field, since we're adding it to the 1.0 version of the protocol.
  This means any 1.0 client can send/receive this message.
* We did not preserve *binary* compatibility of the API, only *protocol* compatibility.  This should be "ok" since server and client
  are completely separated via classloader, and any client will have to pull in the protocol library.
* We did not do extensive testing to make sure the appropriate values are returned in an integration test.  That remains a TODO.

Thanks to @dancingrobot84 for the original patch.

Review by @havocp (Protocol update review by @eed3si9n )

cc> @dancingrobot84 @dpratt

### CODE REVIEW

File: commons/protocol/src/main/scala/sbt/protocol/Keys.scala [hunk 1]
```diff
@@ -82,11 +82,29 @@ object KeyList {
   implicit val pickler: Pickler[KeyList] = genPickler[KeyList]
 }
 
+// TODO - This is a bad name...
```
Comment (havocp) [inline commons/protocol/src/main/scala/sbt/protocol/Keys.scala]: btw I saw https://github.com/naholyr/github-todos on twitter the other day and thought of you :-)

File: commons/protocol/src/main/scala/sbt/protocol/Keys.scala [hunk 2]
```diff
@@ -82,11 +82,29 @@ object KeyList {
...
+  implicit val picklerUnpickler = PicklerUnpickler.generate[ProjectDependencies]
+}
+
 /** Core information returned about projects for build clients. */
 final case class MinimalProjectStructure(
   id: ProjectReference,
   // Class names of plugins used by this project.
-  plugins: Vector[String])
+  plugins: Vector[String],
+  /** A possible list of dependencies between projects..   Added in 0.2 of the protocol. */
```
Comment (havocp) [inline commons/protocol/src/main/scala/sbt/protocol/Keys.scala]: The current version is "1"  not "0.1" iirc. also, it doesn't look like you actually bumped the version ?

It seems a bit odd to bump it on every added field, but Protocol.scala comment actually says that's the plan...

**Comment by havocp** [comment]
looks reasonable but I guess we need to spell out what we do with protocol version for stuff like this. are we bumping it on every patch that adds a field, or only when we go to make a release? and the comment in Protocol.scala says we'll always be back-compat even when we bump it, which I think is right, but then why do we have the protocol-versioned directory in the tests?

File: commons/protocol/src/main/scala/sbt/protocol/Keys.scala (hunk 2 above)
Comment (jsuereth) [inline commons/protocol/src/main/scala/sbt/protocol/Keys.scala]: Yeah.... I actually think maybe we can leave it at version 1 for this.   I'd like to save bumps for major-major changes, but maybe we should just do it... I dunno.

**Comment by havocp** [comment]
I'm fine with saying we'll bump the protocol version essentially never. I think there's almost always a better answer - back-compat addition, feature tags, changing the lockfile we use to launch, etc.

The one argument for bumping it on every added field would be that then clients or servers can fail fast if they truly require a newer field. For example if a client needs this info you've added here and you bump to 2, then it could require 2 on startup and complain otherwise.

File: commons/protocol/src/main/scala/sbt/protocol/Keys.scala (hunk 2 above)
Comment (eed3si9n) [inline commons/protocol/src/main/scala/sbt/protocol/Keys.scala]: SemanticVersioning 2.0.0 Article 7:

> Minor version Y (x.Y.z | x > 0) MUST be incremented if new, backwards compatible functionality is introduced to the public API.

So we need to go 1.1. This allows us to add new batch of oneway trip files to make sure 1.0 input can be parsed using current 1.1 code, and vice versa.

Comment (havocp) [inline commons/protocol/src/main/scala/sbt/protocol/Keys.scala]: protocol version right now isn't necessarily tied to the jar version, fwiw. I don't think it's clear how we want to use the protocol version. at the moment the protocol version is simply an integer, not a semantic version with multiple components. breaking changes in the protocol aren't allowed at all ... conceptually the protocol version is probably the "Y" in the semver but I'm not sure.

Comment (eed3si9n) [inline commons/protocol/src/main/scala/sbt/protocol/Keys.scala]: JAR versions in Scala ecosystem has little meaning, and definitely not Semantic Versioning.
If we are dealing with external users, we should reserve some mechanism to make breaking changes at some point in the future. Is it not possible at this point to expand protocol version to "X.Y.Z" format from plain integer?

**Comment by havocp** [comment]
So here's an attempt to sketch how changes to the protocol get handled:

 * add new fields: bumping the protocol version allows clients to throw an error if they get a too-old server and require the new fields.
 * add new request: client can try it and get an error if not supported, or bumping protocol version allows them to theoretically check in advance
 * conditionally break semantics of existing messages/fields ("feature negotiation"): this can use the tags mechanism. if both client and server have the "such-and-such works in a new way" tag then they both switch over to the new way. alternatively, we could bump protocol version and if both have a certain version the semantics change, or something like that. Details will be worked out by the first person to want to break semantics...
 * unconditionally break semantics of existing messages/fields: we should 1) not do this, but 2) if we do, handle it during the server launch process so that we always launch a server matching the client, OR fail to launch if there's an existing incompatible server. e.g. we could use a distinct lockfile for distinct server generations.

File: commons/protocol/src/main/scala/sbt/protocol/Keys.scala (hunk 2 above)
Comment (havocp) [inline commons/protocol/src/main/scala/sbt/protocol/Keys.scala]: for a full breaking change I think we want to change the launch protocol, so clients don't even connect to the wrong kind of server.

**Comment by jsuereth** [comment]
@havocp @eed3si9n Check the new infrastructure i added to test new protocols and the bump to protoocl version 2.

**Comment by havocp** [comment]
LGTM. will let Eugene also look.

**Comment by eed3si9n** [comment]
LGTM

**Pull request merged**